### PR TITLE
Adds support for idempotent producers

### PR DESCRIPTION
**Pull request by jlerche**

### CODE REVIEW

File: kafkaserver/handler.go [hunk 1]
```diff
@@ -75,7 +75,23 @@ func (n *NewHandler) HandleProduceRequest(_ *protocol.RequestHeader, req *protoc
 				cf.CountDown(nil)
 				continue
 			}
+
 			producedRecords := partitionData.Records[0]
+			producerID := int(binary.BigEndian.Uint64(producedRecords[43:51]))
```
Comment (purplefox) [inline kafkaserver/handler.go:80]: I think we should move this chunk of code to KafkaInOperator.HandleStreamBatch - this is called on the eventloop with the batch. Then you can extract the producer id and sequence there and lookup from the map stored there

File: kafkaserver/handler.go [hunk 2]
```diff
@@ -75,7 +75,23 @@ func (n *NewHandler) HandleProduceRequest(_ *protocol.RequestHeader, req *protoc
...
+			baseSequence := int(binary.BigEndian.Uint32(producedRecords[53:57]))
+			lastOffsetDelta := int(binary.BigEndian.Uint32(producedRecords[23:27]))
+			sequenceNumber := baseSequence + lastOffsetDelta
+			// if producerID is 0 then idempotency is disabled on the producer
+			if producerID > 0 {
+				lastSequenceNumber, exists := topicInfo.ProduceInfoProvider.GetIdempotentProducerMetadata(producerID)
+				if exists && sequenceNumber <= lastSequenceNumber {
+					partitionResponses[j].ErrorCode = protocol.ErrorCodeDuplicateSequenceNumber
```
Comment (purplefox) [inline kafkaserver/handler.go:88]: And will need to return the other error code if there are gaps in the sequence (see KIP)

**Comment by purplefox** [comment]
Hi @jlerche - just wondering how this is going? Do you need any help? :)

**Comment by jlerche** [comment]
> Hi @jlerche - just wondering how this is going? Do you need any help? :)

Should be ready to go! @purplefox

File: opers/kafka_in.go [hunk 3]
```diff
@@ -77,14 +81,53 @@ func (k *KafkaInOperator) IngestBatch(recordBatchBytes []byte, processor proc.Pr
...
+	if producerID > -1 {
+		baseSequence := int(binary.BigEndian.Uint32(bytes[53:57]))
+		lastOffsetDelta := int(binary.BigEndian.Uint32(bytes[23:27]))
+		sequenceNumber := baseSequence + lastOffsetDelta
+		if processorID >= 0 && processorID < len(k.partitionProducerMapping) {
+			if k.partitionProducerMapping[processorID] == nil {
+				k.partitionProducerMapping[processorID] = make(map[int]map[int]int)
+			}
+		} else {
+			panic("unexpected processor ID")
```
Comment (purplefox) [inline opers/kafka_in.go:96]: best to return error not panic

File: opers/kafka_in.go [hunk 4]
```diff
@@ -77,14 +81,53 @@ func (k *KafkaInOperator) IngestBatch(recordBatchBytes []byte, processor proc.Pr
...
+		}
+		if k.partitionProducerMapping[processorID][partitionID] == nil {
```
Comment (purplefox) [inline opers/kafka_in.go:98]: If you store this in a local variable you can avoid the double lookup multiple times

**Review by purplefox — CHANGES_REQUESTED** [review]
Looking good, just a couple of comments.

I think we also need a test which tests it works with an actual client. Perhaps take a look at integration/producer_endpoint_test.go or integration/kafka_api_test.go

**Comment by jlerche** [comment]
> Looking good, just a couple of comments.
> 
> I think we also need a test which tests it works with an actual client. Perhaps take a look at integration/producer_endpoint_test.go or integration/kafka_api_test.go

I enabled idempotence in the go client part of `producer_endpoint_test.go`. I could add it to the java one too and then have them both run their test with and without idempotency, what do you think?

**Comment by purplefox** [comment]
OK thank you!
Perhaps you could. do a quick manual test and stick a breakpoint in your IDE where the dedup code occurs and make sure it is called when running the producer endpoint test?

**Comment by jlerche** [comment]
Sure, here's a screenshot when running the test

image

**Pull request merged**